xdem/coreg.py

```diff
@@ -1442,3 +1482,457 @@ def _to_matrix_func(self) -> np.ndarray:
+                f"\n... and {len(exceptions) - 5} more" if len(exceptions) > 5 else ""
+            )
+
+        # Set the _fit_called parameters (only identical copies of self.coreg have actually been called)
+        self.coreg._fit_called = True
+        if hasattr(self.coreg, "pipeline"):
+            for step in self.coreg.pipeline:
+                step._fit_called = True
+
+    def _restore_metadata(self, meta: dict[str, Any]) -> None:
+        """
+        Given some metadata, set it in the right place.
+
+        :param meta: A metadata file to update self._meta
+        """
+        self.coreg._meta.update(meta)
+
+        if hasattr(self.coreg, "pipeline") and "pipeline" in meta:
+            for i, step in enumerate(self.coreg.pipeline):
+                step._meta.update(meta["pipeline"][i])
+
+    def to_points(self) -> np.ndarray:
+        """
+        Convert the piecewise coregistration matrices to 3D (source -> destination) points.
+
+        The returned shape is (N, 3, 2) where the dimensions represent:
+            0. The point index where N is equal to the amount of subdivisions.
+            1. The X/Y/Z coordinate of the point.
+            2. The old/new position of the point.
+
+        To acquire the first point's original position: points[0, :, 0]
+        To acquire the first point's new position: points[0, :, 1]
+        To acquire the first point's Z difference: points[0, 2, 1] - points[0, 2, 0]
+
+        :returns: An array of 3D source -> destionation points.
+        """
+        if len(self._meta["coreg_meta"]) == 0:
+            raise AssertionError("No coreg results exist. Has '.fit()' been called?")
+        points = np.empty(shape=(0, 3, 2))
+        for meta in self._meta["coreg_meta"]:
+            self._restore_metadata(meta)
+
+            #x_coord, y_coord = rio.transform.xy(meta["transform"], meta["representative_row"], meta["representative_col"])
+            x_coord, y_coord = meta["representative_x"], meta["representative_y"]
+
+
+            old_position = np.reshape([x_coord, y_coord, meta["representative_val"]], (1, 3))
+            new_position = self.coreg.apply_pts(old_position)
+        
+            points = np.append(points, np.dstack((old_position, new_position)), axis=0)
+
+        return points
+
+    def stats(self) -> pd.DataFrame:
+        """
+        Return statistics for each chunk in the piecewise coregistration.
+
+            * center_{x,y,z}: The center coordinate of the chunk in georeferenced units.
+            * {x,y,z}_off: The calculated offset in georeferenced units.
+            * inlier_count: The number of pixels that were inliers in the chunk.
+            * nmad: The NMAD after coregistration.
+            * median: The bias after coregistration.
+
+        :raises ValueError: If no coregistration results exist yet.
+
+        :returns: A dataframe of statistics for each chunk.
+        """
+        points = self.to_points()
```

Comment:
How long does this function take? You're calling it to fit the function, to get the statistics etc. I guess some calculations are done several times.
Is there a way to avoid repeat some of the calculation?

Reply:
It takes four seconds on my laptop. Without multiple threads, it takes about a minute.

Reply:
A minute is quite long when you need to do it multiple times. And that's only for a small DEM...
It's always the drawback when splitting one code into smaller chunks, one might end up needing the same info twice and re-doing the same calculation...
So maybe the stats should be calculated by default and stored, so one does not need to redo this long calculation when one wants to output the stats?

Reply:
Oh wait, sorry I thought you meant the `fit` method!! I haven't benchmarked this part but I think it's more or less instantaneous. I can check after lunch.